### DISCUSSION TITLE
Visible bounds for a* and b* in CIELAB

### DISCUSSION THREAD

**Discussion by blue-j** [colour-science/colour#1236] [General]
I have attempted to compute these with the code below.  have i done anything wrong?  i only find this resource where another person calculated same:

[https://www.russellcottrell.com/photo/visualGamut.htm](https://www.russellcottrell.com/photo/visualGamut.htm)

and we do not agree.

my output is:

a* bounds:  (-296.85093304123535, 223.28403152638191)
b* bounds:  (-177.76534310171704, 167.0584579274007)

```
import numpy as np
import matplotlib.pyplot as plt
import colour


ILLUMINANT_IN_XY = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D65']


def calculate_lab_bounds(wavelength_start=360, wavelength_end=780):
    # Generate wavelengths from the visible spectrum
    wavelengths = np.linspace(wavelength_start, wavelength_end, wavelength_end - wavelength_start + 1)
    # Calculate XYZ values for each wavelength
    XYZ = [colour.wavelength_to_XYZ(wavelength) for wavelength in wavelengths]
    # Convert XYZ to Lab, using D65 and 2° observer
    Lab = [colour.XYZ_to_Lab(xyz, ILLUMINANT_IN_XY) for xyz in XYZ]
    # Extract a* and b* channels
    a_values = [value[1] for value in Lab]
    b_values = [value[2] for value in Lab]

    # Find the min and max for both a* and b*
    a_min, a_max = min(a_values), max(a_values)
    b_min, b_max = min(b_values), max(b_values)

    return (a_min, a_max), (b_min, b_max)

# Use the function to calculate the bounds
a_bounds, b_bounds = calculate_lab_bounds()
print("a* bounds: ", a_bounds)
print("b* bounds: ", b_bounds)

# Plot the a* and b* values across the visible spectrum
wavelengths = np.linspace(360, 780, 421)
a_values, b_values = [], []
for wavelength in wavelengths:
    XYZ = colour.wavelength_to_XYZ(wavelength)
    Lab = colour.XYZ_to_Lab(XYZ)
    a_values.append(Lab[1])
    b_values.append(Lab[2])

plt.figure(figsize=(10, 5))
plt.plot(wavelengths, a_values, 'r', label='a*')
plt.plot(wavelengths, b_values, 'b', label='b*')
plt.title('CIELAB a* and b* Values Across the Visible Spectrum')
plt.xlabel('Wavelength (nm)')
plt.ylabel('CIELAB a* and b* Values')
plt.legend()
plt.grid(True)
plt.show()
```

![visible_spectrum_lab](https://github.com/colour-science/colour/assets/3990142/7ecb2889-408f-4ed5-af6d-6e9e83f91a0b)

**Comment by blue-j** [comment]
I also seem to see that the visible wavelength upper bound is set at 780nm? but the CIE states it to be 830nm, and that is what Mr Cottrell uses as well. (it seems that 360nm as the lower bound works fine in my code though.)

> The joint ISO/CIE International Standard on CIE tristimulus values (ISO/CIE, 2012)
> recommends that the CIE tristimulus values of a colour stimulus should be obtained by
> multiplying at each wavelength the value of the colour stimulus function... by that of each
> of the CIE colour-matching functions, and integrating each set of products over the wavelength
> range corresponding to the entire visible spectrum, **360 nm to 830 nm**.
[CIE 015:2018, pg. 21]

> The wavelength range for standard illuminants is 300 nm to 830 nm because, for evaluating photoluminescent material, the ultraviolet spectral range is relevant. For most colorimetric investigations the restricted wavelength range, 380 nm to 780 nm, can be used.
[CIE 015:2018, pg. 11]

Note also that the CIE (1931) 2-deg color matching functions use the range of 360 - 830nm.

the visible spectrum is not the same as "most colorimetric investigations," i believe...  I am wrong all the time, please educate me if i am, so i can be wrong for as little time as possible!!!

- J

**Reply by thomasmansencal** [reply]
`[360, 830]` is the range typically published by the CIE but we do use the ASTM on in Colour for colorimetric computations, reason being is that they standardised how to do so (ASTME-308) where the CIE did not really to the same level!

As for your original questions, here is what I obtained using CMFS with the `[360, 780, 5]` spectral shape and Illuminant E:

```python
import colour
import numpy as np

XYZ = colour.volume.XYZ_outer_surface()
Lab = colour.XYZ_to_Lab(XYZ, [1/ 3, 1 / 3])

print(np.min(Lab[..., 1]), np.max(Lab[..., 1]))
print(np.min(Lab[..., 2]), np.max(Lab[..., 2]))
```

```
-165.669568905 141.769283507
-132.065827323 146.664895569
```

To visualise the Spectral Locus and the *Rösch-MacAdam* Solid Boundaries (name of the "Visible Gamut") in *CIE Lab*, you could use [colour-visuals](https://github.com/colour-science/colour-visuals):

```python
import colour_visuals
import numpy as np
import pygfx as gfx
from wgpu.gui.auto import WgpuCanvas, run

canvas = WgpuCanvas(size=(960, 540))
renderer = gfx.renderers.WgpuRenderer(canvas)
camera = gfx.PerspectiveCamera(50, 16 / 9)
controller = gfx.OrbitController(camera)
controller.register_events(renderer)

scene = gfx.Scene()
scene.add(gfx.Background(None, gfx.BackgroundMaterial(np.array([0.18, 0.18, 0.18]))))

visuals = [
    colour_visuals.VisualGrid(size=4),
    colour_visuals.VisualSpectralLocus3D(model="CIE Lab"),
    colour_visuals.VisualRoschMacAdam(model="CIE Lab"),
]

group = gfx.Group()
for visual in visuals:
    group.add(visual)

scene.add(group)

camera.local.position = np.array([-0.5, -2.5, 5])
camera.show_pos(np.array([0, 0, 0.5]))

canvas.request_draw(lambda: renderer.render(scene, camera))
run()
```

![image](https://github.com/colour-science/colour/assets/99779/ef4ba3a5-c67d-40fc-9051-ddddb30fed92)

Note that CIE Lab coordinates are normalised and divided by 100 for convenience.

The issue boils down to the incorrect assumption that the Spectral Locus should be on the surface of the *Rösch-MacAdam* Solid Surface. In order to define the boundaries, spectra must be integrated with our sensitivities, i.e. CMFS. This is typically done by using square pulses with growing bandpass. Taken from our documentation:


```reStructuredText
    Assuming 5 bins, a first set of SPDs would be as follows::

        1 0 0 0 0
        0 1 0 0 0
        0 0 1 0 0
        0 0 0 1 0
        0 0 0 0 1

    The second one::

        1 1 0 0 0
        0 1 1 0 0
        0 0 1 1 0
        0 0 0 1 1
        1 0 0 0 1

    The third:

        1 1 1 0 0
        0 1 1 1 0
        0 0 1 1 1
        1 0 0 1 1
        1 1 0 0 1
```
